lib/pbench/server/api/resources/users_api.py

```diff
@@ -417,27 +422,30 @@ def put(self, username):
                 "User trying to update fields that are not present in the user database. Fields: {}",
                 non_existent,
             )
-            abort(400, message="Invalid fields in update request payload")
+            abort(
+                HTTPStatus.BAD_REQUEST,
+                message="Invalid fields in update request payload",
+            )
         protected = set(post_data.keys()).intersection(set(User.get_protected()))
```

Comment:
If the requestor is an admin and the target is the requestor, then I think we should (at least for the interim) add the `role` field to the "protected" set, to prevent the last admin from de-admin'ing him/herself.

Reply:
Okay this means we _can_ delete an admin user via API if the requester user itself is an admin until, do we want to allow that or restrict admin deletion operation via CLI only, since an admin user should have an access to the server.

Reply:
Right now we have no provision for the concept of "an" admin user; we only have username "admin" (and, come to think of it, we need to decide what that looks like if we adopt an "email is username" model). I'd be inclined to say, at least for now, that only the CLI can delete "admin" (whatever it's called) but anyone else can potentially be edited/deleted in the API regardless of the user's role.

Reply:
@dbutenhof, I'm pretty sure that Nikhil has implemented "admin-ness" as a **role**.  Once that's merged, we can have more than one admin user, and they don't have to have a particular name/e-mail.

If you guys are "sure", then I'm willing to accept (at least for the moment) the requirement that an API `DELETE` request is not allowed to delete _any_ user with the "admin" role.  (I concur that the CLI should be allowed to do basically _anything_.)  However, I think it's adequate to allow deletions of _other_ admin users via the API -- we just shouldn't allow an admin to delete itself.

Nevertheless, pursuant to this conversation here, the question is whether an admin should be allowed to remove the "admin" role from itself.  And, I think the answer to that (given the present discussion) is a clear "no".  (Which leaves the question of whether an admin should be allowed to remove the admin role from another user, and I'm gonna wait on asking/answering that question until I hear what the consensus is on whether admin users can delete other admin users....)

Reply:
> Nevertheless, pursuant to this conversation here, the question is whether an admin should be allowed to remove the "admin" role from itself. And, I think the answer to that (given the present discussion) is a clear "no". (Which leaves the question of whether an admin should be allowed to remove the admin role from another user, and I'm gonna wait on asking/answering that question until I hear what the consensus is on whether admin users can delete other admin users....)

I'm not convinced of that. I think we need a standard "admin/admin" user (although I have no idea what to call if it we're using email for usernames) to jumpstart the process. However this can only be created with the CLI, and it can be recreated, so I'm not certain it needs any particular protections. I think an ADMIN (role) user can add or remove the ADMIN role from any user; they have ADMIN because they're trusted. If they delete ADMIN from the last user to have it, they go to the CLI and create a new user with ADMIN, or add it to an existing user; no big deal.

Reply:
I'm inclined to turn your logic on itself:  if someone can just go to the CLI, then they can modify or delete any user they want, and so it doesn't matter that we impose limits on what they can do through the API.  Moreover, we don't need to have a standard "admin/admin" user (or worry about what to call it), because they can just go to the CLI and create it any way they want it.

Given that, I'm inclined to make the API "safe", so that it's hard to make big mistakes using it (and so that it's harder for hostile actors to mess things up).  I'm imagining deployments where someone with an admin role does _not_ necessarily have access to the CLI, in which case having some "blade guards" in place would be a good thing, especially since the API might be accessible from anywhere on the Internet.

Reply:
> I'm inclined to turn your logic on itself: if someone can just go to the CLI, then they can modify or delete any user they want, and so it doesn't matter that we impose limits on what they can do through the API. Moreover, we don't need to have a standard "admin/admin" user (or worry about what to call it), because they can just go to the CLI and create it any way they want it.

They _can_, but I think our install should provide a standard out-of-the-box admin user they can easily use for startup configuration. They can change the password, or even delete it once they've assigned ADMIN role to another user, but many products do this and it provides at very least a standard template for describing configuration. ("Log in as admin/admin and ...")

Reply:
I'm wondering about the usability of this.  If our standard install provides a standard admin user with a well-known password, then sorta the first order of business on deployment would be to go in and change the admin user password...so why not just go with the idea that, if the person doing the deployment wants to have an admin user, they have to go manually create it -- that's pretty much the same level of effort as changing the password for it.

I'm thinking that in the overwhelming number of use-cases, an admin user isn't even required...so why create one by default?  Moreover, if we delegate that to the person doing the deployment, then we finesse all the issues of what the admin username should be, what its e-mail should be, and what it's password should be -- those are all details which the system administrator would have, so we don't need to come up with them!

Reply:
I think this one is addressed, okay to resolve?

Reply:
In the context of this PR, yeah, I think it's resolved.